rfcs/2020-11-02-remap-infallible-functions.md

```diff
@@ -0,0 +1,371 @@
+# RFC … - 2020-11-02 - Remap Language Fallible and Infallible Functions
+
+By introducing a distinction between fallible and infallible functions, we can
+provide more runtime guarantees to users of the Remap language.
+
+## Scope
+
+- The RFC describes how users can call both fallible and infallible variants of
+  functions.
+- It does so through a more broader context of fallible and infallible
+  _expressions_
+- It _does not_ decide on which expressions (other than functions) should
+  support the concept of infallibility.
+
+## Motivation
+
+As we expand the use-case for the Remap language within Vector, we want to be
+able to guide users as much as possible as they have their first experience with
+the language in Vector.
+
+One way to do so is to have programs run _as predictable as possible_. This
+means moving errors as much as possible to boot-time, while reducing the list of
+potential runtime exceptions.
+
+This RFC proposes a way to allow users to determine how they want to handle
+runtime errors for each individual function, so that they can decide if an error
+is truly an exception and should abort program execution, or if the error can
+safely be ignored.
+
+## Internal Proposal
+
+The following Remap program is easy to understand:
+
+```javascript
+.message = parse_json(.message)
+.timestamp = now()
+```
+
+It:
+
+1. Reads the value of the `message` field of the object.
+2. Parses the value to a JSON value (a literal, array, or map).
+3. Assigns the new value to the `message` field, removing the old value.
+4. Assigns the current timestamp to the `timestamp` field.
+
+### Fallible Functions
+
+While the program itself is easy to understand, running it can result in
+unexpected situations.
+
+If whatever value the `message` field contains cannot be parsed as valid JSON,
+the `parse_json` function returns an error.
+
+This results in an outcome where the original event is left untouched, the
+`message` field still contains the original value, and no `timestamp` field
+exists (assuming none existed to start with).
+
+A use-case such as the above is quite common. Fields often need to be parsed as
+JSON values, but at the same time those fields are highly dynamic and might not
+always be considered valid JSON, for example if a bug in an application starts
+generating invalid stringified JSON output.
+
+Users need a way to handle such situations, without having to resort to complex
+solutions.
+
+### Infallible Function Variants
+
+The solution proposed by this RFC is to introduce the concept of _infallible
+function variants_.
+
+An infallible function variant is an exact copy of an existing fallible
+function, but it is guaranteed to never fails.
```

Comment:
Just to clarify, I think we mean "guaranteed by the user to never fail"? It can still fail and will result in the script being aborted if it does.

Reply:
Hmm, I'm not sure if we're on the same page here.

Calling a function as infallible (i.e. without appending the `!` to the call identifier), _will_ guarantee it never fails.

Reply:
I think the terminology might be confusing. By 'fail' I think you mean it mill cause the whole script to abort with an error - or in the case of a condition resolve to false? An infallible one doesn't fail on an error, it returns 'undefined' and the script can choose how to handle this result.